state/txindex/kv/kv.go

```diff
@@ -75,18 +75,30 @@ func (txi *TxIndex) Get(hash []byte) (*types.TxResult, error) {
 	return txResult, nil
 }
 
-// AddBatch indexes a batch of transactions using the given list of tags.
+// AddBatch indexes a batch of transactions using the given list of events. Each
+// key that indexed from the tx's events is a composite of the event type and
+// the respective attribute's key delimited by a "." (eg. "account.number").
+// Any event with an empty type is not indexed.
 func (txi *TxIndex) AddBatch(b *txindex.Batch) error {
 	storeBatch := txi.store.NewBatch()
 	defer storeBatch.Close()
 
 	for _, result := range b.Ops {
 		hash := result.Tx.Hash()
 
-		// index tx by tags
-		for _, tag := range result.Result.Tags {
-			if txi.indexAllTags || cmn.StringInSlice(string(tag.Key), txi.tagsToIndex) {
-				storeBatch.Set(keyForTag(tag, result), hash)
+		// index tx by events
```

Comment:
Looks good, and seems that AND will work over different types as well, correct?

I would love to see an high-level integration test of this (or the pubsub system in general), adding a batch of transactions and subscribing to events (parsed from query strings) to see if they match. This will be the interface uses receive, good to make sure it is useable (especially the mapping from go types to query strings)

Reply:
> Looks good, and seems that AND will work over different types as well, correct?

Yes.

As far as integration tests, is there anything in `types/event_bus_test.go` or `libs/pubsub/pubsub_test.go` that you believe doesn't cover what you expect it to? Both of those closely resemble an integration test(s) in that they create queries (from strings) and subscribe or query for events.